[compute] add `hostname` option to compute instance

hops opened the pull request:
This pull request adds the optional `hostname` option to the compute_instance resource. I've marked this as a draft as I'd like to get your input on a couple of things.

The `computeClient` used in `resourceComputeInstanceV2Read` does not set any microversion and thus the `OS-EXT-SRV-ATTR:hostname` is not returned by the OpenStack API. The `OS-EXT-SRV-ATTR:hostname` is available from microversion 2.3 but until version 2.90 it was only returned to administrator users. That shouldn't be an issue because until microversion 2.90 the hostname could not be specified anyway and was generated from the instance's name.

I'm not sure what the best solution to this problem is. Maybe an additional parameter to pass the microversion which would mean we have to change the calls in `resourceComputeInstanceV2Update` and `resourceComputeInstanceV2ImportState` as well.  
Another important detail is that with microversion 2.47 the response of the flavor changed from `flavor.id` and `flavor.ilnks` to an object with the individual fields of a flavor. However the flavor ID is _not_ available anymore.

I've tested these changes against a DevStack 2025.1 installation and against a real OpenStack environment running 2024.1. On DevStack I ran all acceptance tests beginning with `testAccCheckComputeV2Instance` without errors. The tests on the OpenStack environment were done manually.

* The added tests don't check the hostname (e.g. using `TestCheckResourceAttr`) as it isn't returned by the API.
* data source changes are not included due to the `dataSourceComputeInstanceV2Read` function also not setting any microversion.

kayrus commented:
@hops no worries, this will be addressed in #1899. we can implement the remaining updates later without needing a gophercloud update. my goal was to cover all the caveats at once, although that hadn't been done previously.

> Should I bump the gophercloud version in this pull request as a separate commit or in a separate pull request?

feel free to push it here, but please be prepared to wait longer for tests

hops commented:
@kayrus I rebased my branch and changed your suggestions from the review.
I'm not particular happy with the naming of `isValidHostname` and `validateHostname`, I'm open for alternative ideas.

The update hostname acceptance test only checks if the instance ID hasn't changed as we're still missing `OS-EXT-SRV-ATTR:hostname` due to `resourceComputeInstanceV2Read` not setting the minimum required microversion 2.90. Do you have an idea how on how to best pass the required microversion?

Last but not least, thank you for your fast response time and sorry that I can't keep up with your pace.

hops commented on openstack/resource_openstack_compute_instance_v2_test.go:
```diff
@@ -835,6 +837,86 @@ func TestAccComputeInstanceV2_hypervisorHostname(t *testing.T) {
+				Config: testAccComputeInstanceV2HostnameFqdnConfig(),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckComputeV2InstanceExists(t.Context(), "openstack_compute_instance_v2.instance_1", &instance),
+					// resource.TestCheckResourceAttrPtr("openstack_compute_instance_v2.instance_1", "hostname", instance.Hostname),
```
That won't work as `Hostname` is already a pointer. https://github.com/gophercloud/gophercloud/blob/main/openstack/compute/v2/servers/results.go#L260

kayrus replied:
in `resource.TestCheckResourceAttrPtr("openstack_compute_instance_v2.instance_1", "hostname", instance.Hostname),` you're passing a nil pointer.

in `testAccCheckComputeV2InstanceExists(t.Context(), "openstack_compute_instance_v2.instance_1", &instance),` you're passing a pointer to a struct and then populate this struct new values. the nil get overwritten with a new pointer, but old nil pointer was already passed to a `TestCheckResourceAttrPtr` func.

try this approach:

```diff
diff --git a/openstack/resource_openstack_compute_instance_v2_test.go b/openstack/resource_openstack_compute_instance_v2_test.go
index 7cc7fc87..06b0bced 100644
--- a/openstack/resource_openstack_compute_instance_v2_test.go
+++ b/openstack/resource_openstack_compute_instance_v2_test.go
@@ -910,13 +910,25 @@ func TestAccComputeV2Instance_hostnameFqdn(t *testing.T) {
                                Config: testAccComputeInstanceV2HostnameFqdnConfig(),
                                Check: resource.ComposeTestCheckFunc(
                                        testAccCheckComputeV2InstanceExists(t.Context(), "openstack_compute_instance_v2.instance_1", &instance),
-                                       resource.TestCheckResourceAttrPtr("openstack_compute_instance_v2.instance_1", "hostname", instance.Hostname),
+                                       resource.TestCheckResourceAttrWith("openstack_compute_instance_v2.instance_1", "hostname", testAccCheckComputeV2InstanceHostname(&instance)),
                                ),
                        },
                },
        })
 }
 
+func testAccCheckComputeV2InstanceHostname(instance *servers.Server) func(_ string) error {
+       return func(v string) error {
+               if instance.Hostname == nil {
+                       return fmt.Errorf("empty hostname")
+               }
+               if *instance.Hostname != v {
+                       return fmt.Errorf("unexpected hostname: %s", *instance.Hostname)
+               }
+               return nil
+       }
+}
+
 func testAccCheckComputeV2InstanceDestroy(ctx context.Context) resource.TestCheckFunc {
        return func(s *terraform.State) error {
                config := testAccProvider.Meta().(*Config)
```

hops replied:
Well, that's embarrassing but totally makes sense. Thank you for your guidance!
I've fixed the acceptance tests with your suggestion and they passed on a DevStack 2025.1 environment. I also squashed my commits.

hops commented on openstack/resource_openstack_compute_instance_v2.go:
```diff
@@ -723,26 +752,52 @@ func resourceComputeInstanceV2Read(ctx context.Context, d *schema.ResourceData,
+	if flavorName != "" {
+		flavorID, err = flavorsutils.IDFromName(ctx, computeClient, flavorName)
+		if err != nil {
+			var e gophercloud.ErrResourceNotFound
+			if errors.Is(err, &e) {
```
No `if gophercloud.ResponseCodeIs(err, http.StatusNotFound)` doesn't work because `IDFromName()` doesn't return a `ErrUnexpectedResponsecode` like `flavors.Get()` does.
`IDFromName()` from [gophercloud utils](https://github.com/gophercloud/utils/blob/main/openstack/compute/v2/flavors/utils.go#L20) returns a `ErrResourceNotFound{Name: name, ResourceType: "flavor"}` with the name being the flavor's name.

Generally I like your approach more but have a couple of remarks.
```go
        flavorName := server.Flavor["original_name"].(string)
        flavorID := server.Flavor["id"].(string)
```
This will always cause a type assertion error because either we have the ID (microversion <= 2.46) or the name (microversion >= 2.47) but never both at the same time. I'd suggest to leave it as is (`flavorID, ok := server.Flavor["id"].(string)`) but I'm open to other suggestions.

```go
        if flavorName == "" || flavorID == "" {
                return diag.Errorf("Error setting OpenStack server's flavor: %v", server.Flavor)
        }
```
Same issue as above. Did you mean to write `if flavorName == "" && flavorID == ""`?

Further I think we have to keep the reset of `flavor_id` and `flavor_name` in case the flavor got deleted but the instance is still running. Otherwise it wouldn't behave the same as before this change.
